design/one-pager-package-signature-validation.md

```diff
@@ -0,0 +1,320 @@
+(MRs). Configurations may install Composite Resource Definitions (XRDs) and
+Compositions, which allow users to define abstractions that may compose any
+number of MRs or other Composite Resources (XRs).
+
+Crossplane packages are defined as OCI images, which allows Crossplane users to
+take advantage of existing tooling and distribution channels, as well as the
+myriad of benefits offered by a content-addressable API. Standardizing on this
+ecosystem also allows Crossplane to easily integrate with other projects that
+are built on top of the OCI specification.
+
+An important distinction between running an image on Kubernetes and installing a
+package via Crossplane is that the former invokes the container runtime on a
+Node in the cluster, while the latter fetches the image directly from the
+Crossplane process. Additionally, Crossplane maintains a separate cache of
+package content, allowing it to minimize the number of operations required when
+reconciling a package. Because package content is fetched directly, the
+Crossplane package manager is a convenient location to inject additional logic
+and safeguards. One such feature is image signature verification.
+
+## Goals
+
+- Enable provenance validation of crossplane packages
+- Allow the specification of a trusted public key to use for validation
+- Provide feedback through the status of the package validation through the\
+PackageRevision that triggered the fetch
+
+## Non-Goals
+
+- How to sign the packages
+- Where to store the signatures
+- How to validate policies beyond simple provenance
+
+## Proposal
+Let's start by implementing the minimal viable provenance checking Sigstore
+cosign offers. By making use of a pre-shared public key, we can validate the
+signatures of crossplane packages that were signed by cosign using the paired
+private key.
+
+### Assumptions:
+- The signatures will have to be computed after the crossplane .xpkg is created
+by using the cosign cli. The signatures can be stored alongside the crossplane
+packages in an OCI compliant registry.
+
+- We should perform the validation before the package revision reports a healthy
+status so that it cannot be activated. 
+
+### Implementation
+
+#### Point of Validation
+If we implement validation into the fetcher to evaluate package signature we can
+fail as early as possible in the lifecycle of a package revision.
+```go
+// Fetch fetches a package image.
+func (i *K8sFetcher) Fetch(ctx context.Context, ref name.Reference, secrets ...string) (v1.Image, error) {
+	auth, err := k8schain.New(ctx, i.client, k8schain.Options{
+		Namespace:        i.namespace,
+		ImagePullSecrets: secrets,
+	})
+	if err != nil {
+		return nil, err
+	}
+	return remote.Image(ref, remote.WithAuthFromKeychain(auth), remote.WithTransport(i.transport), remote.WithContext(ctx))
+}
+```
+[Link]: https://github.com/crossplane/crossplane/blob/b01b17353198a8de28664cf1eec601aaaf2fd95a/internal/xpkg/fetch.go#L119
+
+#### Prior Art For Validation:
```

Comment:
The Kyverno image verifications [documentation](https://kyverno.io/docs/writing-policies/verify-images/) could offer some prior art for the configuration of the validation.

Reply:
The ClusterImagePolicies of sigstore policy-controller are also good reference for how we can surface control of which images to validate https://docs.sigstore.dev/policy-controller/overview/#configuring-policy-controller-clusterimagepolicy , however maybe it makes sense for this policy to be more crossplane centric? IE allowing the policy to be directed at all the fields of the crossplane pkg.crossplane.io/v1 crossplane group types?